bokeh/models/sources.py

```diff
@@ -150,16 +155,24 @@ def _data_from_df(df):
 
         '''
         _df = df.copy()
+
+        # Flatten columns
+        if isinstance(df.columns, pd.MultiIndex):
+            _df.columns = ['_'.join(col) for col in _df.columns.values]
```

Comment:
A note: This will work nicely with string names. However would break for integer-like column names.

```
pd.Index([('a', 'b'), ('c', 2), ('c', 1), ('d', 'f')])
```

Which I think, is okay.

Also, in pandas you could use `.map` like 

```
_df.columns = _df.columns.map('_'.join)
```

Reply:
Thanks for your comments. 

> However would break for integer-like column names.

Indeed. I could cast with `str` here, but that might give trouble with other classed names, so it might be better to just only accept strings here implicitly. @bryevdv ?

> Also, in pandas you could use .map like, although a bit less readable perhaps.

I'd rather go for readability here using the list comprehension.

Reply:
AFAIK it is somewhat conventionally discouraged to have pandas column names that are numbers, am I wrong in that impression? if it is, then I agree it's not worth worrying about

Reply:
By convention 'names' are 'strings' I think. I'm 50/50 on just using`str()` and let the user deal with the fallout. I've added a note to the docs on how bokeh expects a flattened frame.

Reply:
@Karel-van-de-Plassche I leave it up to you, I'd only ask that if you change then a new test get added to cover the new case. Either way I'm happy to merge as soon as you indicate `ready`

Reply:
I noticed the non-multiIndex method only accepts strings anyway. I've opted for re-raising the TypeError with a clearer error message instead and added tests for both the MultiIndex and single indexed column cases. If the tests pass, It'll get a ready from me.